### PR TITLE
Version 2.0

### PR DESCRIPTION
**Pull request by empicano**
This PR should implement some changes discussed in #226:

- [x] Remove deprecated connect/disconnect methods
- [x] Remove deprecated `filtered_messages` and `unfiltered_messages` methods
- [x] Switch to client-wide queue
- [x] Rename id to identifier
- [x] Remove deprecated `message_retry_set` Client argument
- [x] Update README and documentation
- [x] Add migration guide

I'm very fond of the idea to switch out `paho` with `rumqtt`, but I think that's better suited for a 3.0, otherwise this gets too big. I have too many ideas and too little time 😄

Feedback is welcome 😋

### CODE REVIEW

**Comment by JonathanPlasse** [comment]
Awesome work.
Will take a look this weekend.

**Comment by frederikaalund** [comment]
Really great to see all the effort put into this :smile: :+1: I'll review this weekend as well. :)

**Review by frederikaalund — APPROVED** [review]
First of all, excellent work on this PR. 👍 It's clearly the result of a thought through process.

Overall, I like the idea of client-wide queue. Seeing the resulting (much simpler) code it looks like a much better design. I also like the idea that our users can build a "distributor" on top of it. It *is* a breaking change and there *will* be push back from our 1.x users (that's always the case when you break something). Still, I think we should move forward with it. We gain a lot in terms of maintainability (the client is much simpler now).

I read through all the code but frankly I don't have much to add. Just two nit-picky comments. That's a testament to the quality of your design and implementation. Again, excellent job on that front! 😄 

I also like the idea to push all the other changes that we have discussed during the past year to a 3.0 release. This way, we can land the 2.0 release right away.

Lastly, it's great to see that you touch the entire code base: Client, documentation, and tests. It's very well put together and an exemplary PR. 👍

File: docs/migration-guide-v2.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+```python
+import asyncio
+import aiomqtt
+
+
+async def main():
+    client = aiomqtt.Client("test.mosquitto.org")
+    await client.__aenter__()
+    await client.publish("temperature/outside", payload=28.4)
+    await client.__aexit__(None, None, None)
```
Comment (JonathanPlasse) [inline docs/migration-guide-v2.md:40]: Maybe use a try except block as suggested in `connecting-to-the-broker.md`.

**Comment by JonathanPlasse** [comment]
Great work! Everything seems good to me.

File: aiomqtt/client.py [hunk 2]
```diff
@@ -340,11 +179,8 @@ def __init__(  # noqa: C901, PLR0912, PLR0913, PLR0915
         username: str | None = None,
         password: str | None = None,
         logger: logging.Logger | None = None,
-        client_id: str | None = None,
-        tls_context: ssl.SSLContext | None = None,
-        tls_params: TLSParameters | None = None,
-        tls_insecure: bool | None = None,
-        proxy: ProxySettings | None = None,
+        identifier: str | None = None,
+        queue_type: type[asyncio.Queue[Message]] = asyncio.Queue,
```
[the remark this answers is not in the document]
Comment (empicano) [reply aiomqtt/client.py]: I'm not very experienced with typing. Do you know how to annotate this by any chance? The changes of my last commit make the Python 3.8 tests fail with `TypeError: 'type' object is not subscriptable`. I can ignore these in different ways, but I'm wondering if there's a cleaner way 😄

**Comment by empicano** [comment]
Thank you both for your valuable comments! 😊

There's another thing that I was not 100% sure about, maybe you have some ideas on that: I currently implemented the messages generator as `Client.messages` _attribute_. We could also implement it as `Client.messages()` _method_ by accessing the current `_messages()` directly rather than calling and assigning it inside `__init__`.

I'm not sure if this is just a question of style or if there are other reasons for or against. I opted for `Client.messages` because that makes it feel more intuitive (to me) that only one generator can be used concurrently.

What do you think?

File: docs/migration-guide-v2.md (hunk 1 above)
Comment (empicano) [reply docs/migration-guide-v2.md]: Good idea, I put `__aenter__` outside the `try` clause, as in [PEP 343](https://peps.python.org/pep-0343/), I hope that gets it right 😅

File: aiomqtt/client.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (JonathanPlasse) [reply aiomqtt/client.py]: Python 3.8 does not implement PEP 585, so you cannot use `asyncio.Queue[Message]` in runtime code.
You can cast the type to fix the error.
```python
queue_type = cast("type[asyncio.Queue[Message]]", asyncio.Queue)
```

**Comment by JonathanPlasse** [comment]
`Client.messages` makes sense for me.

**Comment by frederikaalund** [comment]
> Thank you both for your valuable comments! 😊
> 
> There's another thing that I was not 100% sure about, maybe you have some ideas on that: I currently implemented the messages generator as `Client.messages` _attribute_. We could also implement it as `Client.messages()` _method_ by accessing the current `_messages()` directly rather than calling and assigning it inside `__init__`.
> 
> I'm not sure if this is just a question of style or if there are other reasons for or against. I opted for `Client.messages` because that makes it feel more intuitive (to me) that only one generator can be used concurrently.
> 
> What do you think?

I think it is easiest for our users to have `Client.messages` (as a property). However, this approach comes with some caveats. If two consumers iterate through `Client.message` concurrently then:
 * Each gets unique messages (there is no copying)
 * There is (to my knowledge) no [fairness](https://en.wikipedia.org/wiki/Fair_queuing) guarantees. That is, one consumer may get all the messages.
 * The [routing](https://en.wikipedia.org/wiki/Routing#Delivery_schemes) model is essentially [anycast](https://en.wikipedia.org/wiki/Anycast).
 * If our end users want a different routing model (e.g., [broadcast](https://en.wikipedia.org/wiki/Broadcasting_(networking))) then they have to build that themselves on top of aiomqtt.

The above may be surprising to our users. The easiest "work around" is just to write it very explicitly in the documentation. I don't think we should try to solve this routing problem. That's too difficult.

File: aiomqtt/client.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (empicano) [reply aiomqtt/client.py]: Perfect, thank you!

**Comment by empicano** [comment]
I agree that we shouldn't implement different routing models, that would probably build back the same complexity we had before :+1: Currently, something like this:

```python
import asyncio
import aiomqtt


async def handle(client):
    async for message in client.messages:
        print(message.payload)


async def main():
    client = aiomqtt.Client("test.mosquitto.org")
    async with asyncio.TaskGroup() as tg:
        async with client:
            tg.create_task(handle(client))
            tg.create_task(handle(client))


asyncio.run(main())
```

fails with `RuntimeError: anext(): asynchronous generator is already running`; I think that's a good default.

Thanks for the great feedback! I'll merge and release this then 😊

File: aiomqtt/client.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (JonathanPlasse) [reply aiomqtt/client.py]: You are welcome

**Pull request merged**